<?xml version="1.0"?>
<launch>
    <arg name="world" default="$(find franka_simulation)/world/challenge_simulation.world" />
    <arg name="x" default="0.0" />
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.975" />
    <arg name="moveit_controller_manager" default="fake"/>
    <arg name="arm_id" default="panda"/>
    <arg name="initial_joint_positions" default="-J $(arg arm_id)_joint1 0.00000 
                                                 -J $(arg arm_id)_joint2 -0.733038
                                                 -J $(arg arm_id)_joint3 0
                                                 -J $(arg arm_id)_joint4 -1.98968
                                                 -J $(arg arm_id)_joint5 0
                                                 -J $(arg arm_id)_joint6 1.309
                                                 -J $(arg arm_id)_joint7 0.767945
                                                 -J $(arg arm_id)_finger_joint1 0.04
                                                 -J $(arg arm_id)_finger_joint2 0.04"
    />
    <param name="/use_sim_time" value="true"/>

    <rosparam command="load" file="$(find project)/config/pick_and_place_params.yaml" 
              ns="/pick_and_place_config" />

    <include file="$(find panda_moveit_config)/launch/demo_gazebo.launch" 
             pass_all_args="true"/>

    <node pkg="franka_simulation" type="world_challenge.py" name="world_spawner_rviz"/>
    <node pkg="gazebo_grasp_plugin_ros" type="grasp_event_republisher" name="grasp_publisher" output="screen"/>

    <param name="wp_numbers" type="int" value="5"/>
    <node pkg="franka_core" type="wp_generator.py" name="wp_spawner"/>
    <node pkg="franka_core" type="eval_wp.py" name="wp_tracker"/>

    <node name="spawn_aruco_cube_gazebo" pkg="gazebo_ros" type="spawn_model"
        args="-file $(env HOME)/.gazebo/models/aruco_cube/model.sdf -sdf -model aruco_cube -x 0.5 -y -0.1 -z 1.03"
        output="screen" />
    
    <node pkg="project" type="spawn_red_cube.py" name="spawn_red_cube" output="screen" 
        launch-prefix="bash -c 'sleep 3.0; $0 $@'"/>
    <node pkg="project" type="spawn_green_cube.py" name="spawn_green_cube" output="screen"
        launch-prefix="bash -c 'sleep 3.0; $0 $@'"/>
    <node pkg="project" type="spawn_blue_cube.py" name="spawn_blue_cube" output="screen"
        launch-prefix="bash -c 'sleep 3.0; $0 $@'"/>

    <node pkg="project" type="pick_and_place_core.py" name="pick_and_place_core_node" output="screen"
        launch-prefix="bash -c 'sleep 5.0; $0 $@'"/>
    
    <node pkg="project" type="color_perception.py" name="color_perception_node" output="screen"
        launch-prefix="bash -c 'sleep 7.0; $0 $@'"/>
    
    <node pkg="project" type="aruco_perception.py" name="aruco_perception_node" output="screen"
        launch-prefix="bash -c 'sleep 20.0; $0 $@'"/>

</launch>